### PR TITLE
Add Universal React routing support

### PR DESCRIPTION
**Pull request by ericelliott**
This update adds a `component routes` option to Keystone's config which allows users to pass `react-router` routes into Keystone. In response, Keystone will wire up the routes on the server.

The same route modules can be used for rendering on the client side, as well, which adds universal routing support to Keystone.

You can test this PR using the [`universal-keystone-boilerplate`](https://github.com/ericelliott/universal-keystone-boilerplate).

Clone, configure, `npm install` and run `npm test` to see it working.

You can also load `localhost:3000/teams` in the browser to see a route rendered by the router.

### CODE REVIEW

File: server/createComponentRouter.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+var match = ReactRouter.match;
+var RoutingContext = ReactRouter.RoutingContext;
+
+module.exports = function createComponentRouter (routes) {
+  return function componentRouter (req, res, next) {
+    match({ routes: routes, location: req.url },
+        function (error, redirectLocation, renderProps) {
+
+      if (error) return res.status(500).send(error.message);
```
Comment (ericelliott) [inline server/createComponentRouter.js]: Is there a way to trigger a custom 500 error message from Keystone?

Comment (JedWatson) [inline server/createComponentRouter.js]: what do you mean?

**Comment by JedWatson** [comment]
Looks great @ericelliott, one thing I'd suggest changing: maybe call the setting `react routes` rather than `component routes`, which will make the "React" part emphasised for people new to Keystone. It's still not perfect but the clearest I've been able to come up with... Other suggestions / arguments welcome.

File: server/createComponentRouter.js (hunk 1 above)
Comment (JedWatson) [inline server/createComponentRouter.js]: Also are there missed tabs & spaces in this file? the indentation in the snippet above looks weird

**Comment by JedWatson** [comment]
Maybe `universal react routes`? baah I hate naming settings

**Comment by ericelliott** [comment]
`react routes` sounds good to me. I'll update. =)

File: server/createComponentRouter.js (hunk 1 above)
Comment (ericelliott) [inline server/createComponentRouter.js]: > what do you mean?

I assume Keystone supports rendering custom error messages -- the line above will send a pretty generic error.

> Also are there mixed tabs & spaces in this file? the indentation in the snippet above looks weird

Maybe, but lint passes. I'll check it out. Is there a lint rule to cover that? And did `.editorconfig` fall asleep at the wheel?

**Comment by ericelliott** [comment]
What's the process for updating the docs?

**Comment by webteckie** [comment]
@ericelliott docs are in https://github.com/keystonejs/keystonejs-site

File: server/createComponentRouter.js (hunk 1 above)
Comment (ericelliott) [inline server/createComponentRouter.js]: > Also are there mixed tabs & spaces in this file? the indentation in the snippet above looks weird

It appears there's no mixed spaces and tabs.. but you're right. I'll fix the indentation.

**Comment by ericelliott** [comment]
Updated.

**Comment by ericelliott** [comment]
Super excited about this. Just sayin'.

**Comment by JedWatson** [comment]
me too :grinning:

**Pull request merged**